### ISSUE TITLE
Type narrowing on match statements with dataclass params

### ISSUE THREAD

**Issue by thejaminator** [microsoft/pyright#4942]
**Is your feature request related to a problem? Please describe.**
Hi Eric, first of all I would like to thank you for all your work on pyright, I think its really great.
I'm wondering if we could add type narrowing to the params of dataclasses

```py
from dataclasses import dataclass
from typing import Optional


@dataclass
class UserDataclass:
    email: Optional[str]

def log_user_dataclass(user: UserDataclass) -> str:
    match user:
        case UserDataclass(email=None):
            return ""
        case UserDataclass(email=email):
            reveal_type(email)
            return email

```

You'll get
```
-  information: Type of "email" is "str | None"
 - error: Expression of type "str | None" cannot be assigned to return type "str"
    Type "str | None" cannot be assigned to type "str"
      Type "None" cannot be assigned to type "str" (reportGeneralTypeIssues)
```
**Describe the solution you'd like**

Since we've matched on `case UserDataclass(email=None):` in the first statement, email should be narrowed to not be None in second case statement.

This is something that already happens in arguments that aren't in a dataclass - e.g.
```py
def log_optional(something: Optional[str]) -> str:
    match something:
        case None:
            return ""
        case something:
            reveal_type(something)
            return something
```

**Additional context**
Add any other context or screenshots about the feature request here:

**Comment by erictraut** [comment]
The type of `something` is `UserDataclass`. It's type cannot be further narrowed. There's no such thing in the Python type system as "a UserDataclass whose `email` attribute is not a `None`".

In your second example, the type of `something` is `str | None`, so it can be narrowed to `None` or `str`.

If you want a static type checker to handle narrowing in the first code sample, you could use a nested `match` statement:

```python
def log_user_dataclass(user: UserDataclass) -> str:
    match user:
        case UserDataclass(email):
            match email:
                case None:
                    return ""
                case x:
                    reveal_type(x)
                    return x
```

Or this would work:

```python
def log_user_dataclass(user: UserDataclass) -> str:
    match user:
        case UserDataclass(email):
            return email if email is not None else ""
```

**Comment by thejaminator** [comment]
thanks eric for the reply!
